### PR TITLE
update_db_binary: adapt relocated scylla binary

### PR DESCRIPTION
**Pull request by amoskong**
Scylla binary is moved to different directory after relocation,
then we can't always copy binary to original /usr/bin/scylla.

### CODE REVIEW

**Comment by amoskong** [comment]
scylla-3.0.3-0.20190213.873e0f0e1.el7.x86_64.rpm: not relocated 
scylla-3.0.3-0.20190221.35cc09b15.el7.x86_64.rpm: relocated

File: sdcm/cluster.py [hunk 1]
```diff
@@ -1423,7 +1423,7 @@ def start_scylla_jmx(self, verify_up=True, verify_down=False, timeout=300):
     @log_run_info
     def start_scylla(self, verify_up=True, verify_down=False, timeout=300):
         self.start_scylla_server(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
-        self.start_scylla_jmx(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
+        self.start_scylla_jmx(verify_up=verify_up, verify_down=False, timeout=timeout)
```
Comment (roydahan) [inline sdcm/cluster.py]: What is the reason for this change?

File: sdcm/cluster.py [hunk 2]
```diff
@@ -1448,7 +1448,7 @@ def stop_scylla_jmx(self, verify_up=False, verify_down=True, timeout=300):
     @log_run_info
     def stop_scylla(self, verify_up=False, verify_down=True, timeout=300):
         self.stop_scylla_server(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
-        self.stop_scylla_jmx(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
+        self.stop_scylla_jmx(verify_up=False, verify_down=verify_down, timeout=timeout)
```
Comment (roydahan) [inline sdcm/cluster.py]: Same question as above

File: sdcm/cluster.py (hunk 1 above)
Comment (roydahan) [reply sdcm/cluster.py]: I see now the commit message.
Did you touch this issue describe in the commit message?

Comment (amoskong) [reply sdcm/cluster.py]: When I tested my first first update_db_binary patch,  it failed to stop non-seed nodes (timeout).

In the sct log I found scylla-server had been stop at that point, then try to stop scylla-jmx. The code will try to verify jmx is up before stop it.

```
1448     @log_run_info
1449     def stop_scylla(self, verify_up=False, verify_down=True, timeout=300):
1450         self.stop_scylla_server(verify_up=verify_up, verify_down=verify_down, timeout=timeout)  << scylla-server had been stop, jmx was also stop for dependency
1451         self.stop_scylla_jmx(verify_up=verify_up, verify_down=verify_down, timeout=timeout)   <<< verify_up is also True here

```

```
1840     def _update_db_binary(self, new_scylla_bin, node_list):
1841         self.log.debug('User requested to update DB binary...')
1842 
1843         seed_nodes = self.get_seed_nodes()
1844         non_seed_nodes = [n for n in self.nodes if not n.is_seed]
1845 
1846         def update_scylla_bin(node, queue):
1847             node.log.info('Updating DB binary')
1848             node.remoter.send_files(new_scylla_bin, '/tmp/scylla', verbose=True)
1849             # replace the binary
1850             node.remoter.run('sudo cp -f /usr/bin/scylla /usr/bin/scylla.origin')
1851             node.remoter.run('sudo cp -f /tmp/scylla /usr/bin/scylla')
1852             node.remoter.run('sudo chown root.root /usr/bin/scylla')
1853             node.remoter.run('sudo chmod +x  /usr/bin/scylla')
1854             queue.put(node)
1855             queue.task_done()
1856 
1857         def stop_scylla(node, queue):
1858             node.stop_scylla(verify_down=True, verify_up=True)   <<<   We assigned verify_down to True here
1859             queue.put(node)
1860             queue.task_done()
1861 
1862         def start_scylla(node, queue):
1863             node.start_scylla(verify_down=True, verify_up=True)
1864             queue.put(node)
1865             queue.task_done()
1866 
1867         start_time = time.time()
1868 
1869         # First, stop *all* non seed nodes
1870         self.run_func_parallel(func=stop_scylla, node_list=non_seed_nodes)   <<< failed top stop scylla node
1871         # First, stop *all* seed nodes
1872         self.run_func_parallel(func=stop_scylla, node_list=seed_nodes)
1873         # Then, update bin only on requested nodes
1874         self.run_func_parallel(func=update_scylla_bin, node_list=node_list)

```

```
1438     def stop_scylla_jmx(self, verify_up=False, verify_down=True, timeout=300):
1439         if verify_up:
1440             self.wait_jmx_up(timeout=timeout)  <<< TIMEOUT occurred here!!!
1441         if not self.is_ubuntu14():
1442             self.remoter.run('sudo systemctl stop scylla-jmx.service', timeout=timeout)
1443         else:
1444             self.remoter.run('sudo service scylla-jmx stop', timeout=timeout)
1445         if verify_down:
1446             self.wait_jmx_down(timeout=timeout)
```

File: sdcm/cluster.py (hunk 2 above)
Comment (amoskong) [reply sdcm/cluster.py]: My comment (https://github.com/scylladb/scylla-cluster-tests/pull/807/commits/f505bd2ffa8a0342a3e18089730ba3ddd554587a#r259655566) is about stop_scylla()

File: sdcm/cluster.py (hunk 1 above)
Comment (amoskong) [reply sdcm/cluster.py]: Sorry I above replay is about stop_scylla().  A similar problem exists in start_scylla().

```
1424     def start_scylla(self, verify_up=True, verify_down=False, timeout=300):
1425         self.start_scylla_server(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
1426         self.start_scylla_jmx(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
```

When start_scylla() is called, scylla-server will be start first by line 1425, jmx will also be started for dependency.
In line 1426, it will try to verify jmx is down before trying to start jmx. Actually jmx is already up, then timeout will occur.

Comment (roydahan) [reply sdcm/cluster.py]: So maybe we should completely remove stop/start jmx?
It’s a dependency and we never start/stop it independently.

Comment (amoskong) [reply sdcm/cluster.py]: Hi @bentsi , What's your opinion?

```
diff --git a/sdcm/cluster.py b/sdcm/cluster.py
index 425bfd4..52f6cd6 100644
--- a/sdcm/cluster.py
+++ b/sdcm/cluster.py
@@ -1423,7 +1423,8 @@ server_encryption_options:
     @log_run_info
     def start_scylla(self, verify_up=True, verify_down=False, timeout=300):
         self.start_scylla_server(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
-        self.start_scylla_jmx(verify_up=verify_up, verify_down=False, timeout=timeout)
+        if verify_up:
+            self.wait_jmx_up(timeout=timeout)
 
     def stop_scylla_server(self, verify_up=False, verify_down=True, timeout=300, ignore_status=False):
         if verify_up:
@@ -1448,7 +1449,8 @@ server_encryption_options:
     @log_run_info
     def stop_scylla(self, verify_up=False, verify_down=True, timeout=300):
         self.stop_scylla_server(verify_up=verify_up, verify_down=verify_down, timeout=timeout)
-        self.stop_scylla_jmx(verify_up=False, verify_down=verify_down, timeout=timeout)
+        if verify_down:
+            self.wait_jmx_down(timeout=timeout)
 
     def enable_client_encrypt(self):
         SCYLLA_YAML_PATH_TMP = "/tmp/scylla.yaml"
```

Comment (amoskong) [reply sdcm/cluster.py]: I just append a patch of above fix, please help to review.

File: sdcm/cluster.py [hunk 3]
```diff
@@ -1846,11 +1848,22 @@ def _update_db_binary(self, new_scylla_bin, node_list):
...
+                binary_path = relocated_binary
+            else:
+                binary_path = '/usr/bin/scylla'
             # replace the binary
-            node.remoter.run('sudo cp -f /usr/bin/scylla /usr/bin/scylla.origin')
-            node.remoter.run('sudo cp -f /tmp/scylla /usr/bin/scylla')
-            node.remoter.run('sudo chown root.root /usr/bin/scylla')
-            node.remoter.run('sudo chmod +x  /usr/bin/scylla')
+            prereqs_script = dedent("""
+            cp -f {binary_path} {binary_path}.origin
```
Comment (bentsi) [inline sdcm/cluster.py]: Add a tab to all of the script to improve readability

**Comment by bentsi** [comment]
looks good for me, just improve indent in the `prereqs_script`

**Comment by amoskong** [comment]
Appended a patch to fix indent in the `prereqs_script`, thanks.

File: sdcm/cluster.py (hunk 3 above)
Comment (amoskong) [reply sdcm/cluster.py]: Fixed.

**Comment by bentsi** [comment]
@amoskong BTW, you added a backport tag, but which version to backport to?

**Comment by amoskong** [comment]
3.0

Bentsi <notifications@github.com> 于 2019年2月25日周一 下午11:26写道：
